zap/zap.go

```diff
@@ -14,89 +15,173 @@ var (
 }
 
 // Println adds a log entry with arguments handled in the manner of fmt.Println
-func (zl *Logger) Println(args ...any) {
-	zl.print(fmt.Sprintln(args...))
+func (zpl *Logger) Println(args ...any) {
+	zpl.print(fmt.Sprintln(args...))
 }
 
 // Printf adds a log entry with arguments handled in the manner of fmt.Printf
-func (zl *Logger) Printf(format string, args ...any) {
-	zl.print(fmt.Sprintf(format, args...))
+func (zpl *Logger) Printf(format string, args ...any) {
+	zpl.print(fmt.Sprintf(format, args...))
 }
 
-func (zl *Logger) print(string) {}
+func (zpl *Logger) print(msg string) {
+	msg = strings.TrimSpace(msg)
+	if ce := zpl.logger.Check(zpl.cfg.Level.Level(), msg); ce != nil {
+		ce.Write()
+	}
+
+}
 
 // Debug returns a new logger set to add entries as level Debug
-func (zl *Logger) Debug() slog.Logger {
-	return zl.WithLevel(slog.Debug)
+func (zpl *Logger) Debug() slog.Logger {
+	return zpl.WithLevel(slog.Debug)
 }
 
 // Info returns a new logger set to add entries as level Info
-func (zl *Logger) Info() slog.Logger {
-	return zl.WithLevel(slog.Info)
+func (zpl *Logger) Info() slog.Logger {
+	return zpl.WithLevel(slog.Info)
 }
 
 // Warn returns a new logger set to add entries as level Warn
-func (zl *Logger) Warn() slog.Logger {
-	return zl.WithLevel(slog.Warn)
+func (zpl *Logger) Warn() slog.Logger {
+	return zpl.WithLevel(slog.Warn)
 }
 
 // Error returns a new logger set to add entries as level Error
-func (zl *Logger) Error() slog.Logger {
-	return zl.WithLevel(slog.Error)
+func (zpl *Logger) Error() slog.Logger {
+	return zpl.WithLevel(slog.Error)
 }
 
 // Fatal returns a new logger set to add entries as level Fatal
-func (zl *Logger) Fatal() slog.Logger {
-	return zl.WithLevel(slog.Fatal)
+func (zpl *Logger) Fatal() slog.Logger {
+	return zpl.WithLevel(slog.Fatal)
 }
 
 // WithLevel returns a new logger set to add entries to the specified level
-func (zl *Logger) WithLevel(level slog.LogLevel) slog.Logger {
-	return zl
+func (zpl *Logger) WithLevel(level slog.LogLevel) slog.Logger {
+
+	if zpl.cfg != nil {
+
+		var levels = []zapcore.Level{
+			slog.UndefinedLevel: zapcore.InvalidLevel,
+			slog.Fatal:          zapcore.FatalLevel,
+			slog.Error:          zapcore.ErrorLevel,
+			slog.Warn:           zapcore.WarnLevel,
+			slog.Info:           zapcore.InfoLevel,
+			slog.Debug:          zapcore.DebugLevel,
+		}
+
+		if level < -1 || int(level) >= len(levels) {
+			// fix your code
+			err := fmt.Errorf("slog: invalid log level %v", level)
+			zpl.logger.Panic(err.Error())
+
+		} else if zpl.logger.Core().Enabled(levels[level]) {
+			zpllevel := levels[level]
+			// we set cfg level here
+			zpl.cfg.Level.SetLevel(zpllevel)
+		}
+	}
+
+	return zpl
 }
 
 // WithStack attaches a call stack to a new logger
-func (zl *Logger) WithStack(skip int) slog.Logger {
-	return zl
+func (zpl *Logger) WithStack(skip int) slog.Logger {
+	zpl.logger = zpl.logger.WithOptions(zap.AddStacktrace(zapcore.InfoLevel))
```

Comment:
why `Info` instead of the level chosen by `WithLevel()`?

Reply:
WithStack will enable stack trace for all levels, so I chose the absolute minimal level